### DISCUSSION TITLE
How to prevent leaving the current page when the user manually changes the URL in the browser's address bar?

### DISCUSSION THREAD

**Discussion by PeterM0101** [vercel/next.js#50700] [App Router]
To avoid losing unsaved data, I need to prevent exiting the current page when the user manually changes the URL in the browser's address bar. I'm trying to use the "routeChangeStart" event, which only works fine for switching routes using NextJS tools like Link, router.push, etc. I'm using the "popstate" event, which only works with the browser's back and forward buttons. Could you suggest what event I can use to handle the URL change manually?

**Comment by PeterM0101** [comment]
Solved

**Reply by casperiv0** [reply]
@PeterM0101, how did you solve this issue :)?

**Reply by powerfulyang** [reply]
https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event

**Comment by danielfrg** [comment]
For future reference this is what I ended up with

```javascript
const useLeavePageConfirm = (active = true) => {
  const beforeUnloadListener = (event: any) => {
    event.preventDefault()
    return (event.returnValue = "")
  }

  React.useEffect(() => {
    if (active) {
      addEventListener("beforeunload", beforeUnloadListener)
    } else {
      removeEventListener("beforeunload", beforeUnloadListener)
    }
  }, [active])
}

export default useLeavePageConfirm
```

**Reply by thinhbg2812** [reply]
this solution only work when it's hard reload page, Nextjs use soft navigate cause it not fire when user change page or using back function of browser

**Comment by cgfeel** [comment]
## Intercept route changes at the NextJS app router mode

Demo：[[codeSondbox](https://codesandbox.io/p/devbox/next-block-leaving-yf9nlq?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clq1ovh9a000b3b6lujkvkri7%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clq1ovh9900023b6l4kn66bas%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clq1ovh9900083b6lm7e16teb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clq1ovh99000a3b6lqm95cmz6%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clq1ovh9900023b6l4kn66bas%2522%253A%257B%2522id%2522%253A%2522clq1ovh9900023b6l4kn66bas%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq1p4cpn00023b6ld60iy8s9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A4%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A4%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fapp%252Fform.css%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clq1p4cpn00023b6ld60iy8s9%2522%257D%252C%2522clq1ovh99000a3b6lqm95cmz6%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq1ovh9900093b6l2s8mq53w%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252Fproxy%252Fstore%2522%257D%255D%252C%2522id%2522%253A%2522clq1ovh99000a3b6lqm95cmz6%2522%252C%2522activeTabId%2522%253A%2522clq1ovh9900093b6l2s8mq53w%2522%257D%252C%2522clq1ovh9900083b6lm7e16teb%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq1ovh9900033b6lp2vrvqp7%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clq1ovh9900043b6ll8bpphgd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522build%2522%257D%252C%257B%2522id%2522%253A%2522clq1ovh9900053b6lxnemchtd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%257D%252C%257B%2522id%2522%253A%2522clq1ovh9900063b6lqbse2rr4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522lint%2522%257D%252C%257B%2522id%2522%253A%2522clq1ovh9900073b6lobmjoxn5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522install%2522%257D%255D%252C%2522id%2522%253A%2522clq1ovh9900083b6lm7e16teb%2522%252C%2522activeTabId%2522%253A%2522clq1ovh9900033b6lp2vrvqp7%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)]

https://github.com/cgfeel/next.v2/assets/578141/fe1e7f24-054a-4f56-892c-5345ac177a75

source code: https://github.com/cgfeel/next.v2/tree/master/routing-file/src/app/leaving/proxy

Use this Provider in your layout at the app root directory：https://github.com/cgfeel/next.v2/blob/master/routing-file/src/components/proxyProvider/index.tsx

```tsx
'use client'

import { usePathname, useSearchParams } from "next/navigation";
import Script from "next/script";
import { FC, PropsWithChildren, createContext, useEffect, useState } from "react";

const ProxyContext = createContext<ProxyInstance>([undefined, () => {}]);

const ProxyProvider: FC<PropsWithChildren<{}>> = ({ children }) => {
    const [tips, setTips] = useState<string|undefined>();
    const msg = tips === undefined ? tips : (tips||'Are you sure want to leave this page?');

    const pathname = usePathname();
    const searchParams = useSearchParams();

    const url = [pathname, searchParams].filter(i => i).join('?');
    useEffect(() => {
        setTips(undefined);
    }, [url, setTips]);

    useEffect(() => {
        const handleBeforeUnload = (event: BeforeUnloadEvent) => {
            if (msg === undefined) return msg;
            
            event.preventDefault();
            event.returnValue = msg;
            
            return msg;
        };

        const script = document.getElementById('proxy-script');
        if (script) {
            script.dataset.msg = msg||'';
            script.dataset.href = location.href;
        }

        window.addEventListener("beforeunload", handleBeforeUnload);
        return () => {
            window.removeEventListener("beforeunload", handleBeforeUnload);
        }
    }, [msg]);

    return (
        <ProxyContext.Provider
            value={[msg, setTips]}
        >
            <Script
                strategy="afterInteractive"
                id="proxy-script"
                dangerouslySetInnerHTML={{
                    __html: `(() => {
                        const originalPushState = history.pushState.bind(history);
                        let currentPoint = 0;
                        let point = 0;
                        window.history.pushState = function(state, title, url) {
                            state.point = ++point;
                            currentPoint = point;
                            originalPushState(state, title, url);
                        };
                        const originalReplaceState = history.replaceState.bind(history);
                        window.history.replaceState = function(state, title, url) {
                            state.point = currentPoint;
                            originalReplaceState(state, title, url);
                        };
                        window.addEventListener('popstate', function (event) {
                            const { state: nextState } = event;
                            const isback = currentPoint > nextState.point;

                            currentPoint = nextState.point;

                            const script = document.getElementById('proxy-script');
                            if (!script || location.href === script.dataset.href) return;

                            const msg = script.dataset.msg||'';
                            const confirm = msg == '' ? true : window.confirm(msg);
                            if (!confirm) {
                                event.stopImmediatePropagation();
                                isback ? history.forward() : history.back();
                            }
                        });
                    })()`,
                }}
            ></Script>
            {children}
        </ProxyContext.Provider>
    );
};

export type ProxyInstance = [
    string|undefined, (tips?: string) => void
]

export { ProxyContext };

export default ProxyProvider;
```

**Reply by thinhbg2812** [reply]
im not fully understand the concept of the proxy tbh

**Comment by ab-etemadi** [comment]
## Preventing form page exit in Nextjs 14
I managed to come up with a model that addresses three of these possibilities:
1 - Reload the page (Handling by the browser itself)
2 - Click the “Back” button on the browser (Under our control)
3 - Navigation through application links (Under our control)


Now our Custom hook that will prevent navigation:
```ts
import { useEffect } from 'react';
import { useRouter } from 'next/navigation';
import { NavigateOptions } from 'next/dist/shared/lib/app-router-context.shared-runtime';

export const useWarnIfUnsavedChanges = (unsaved: boolean) => {
  const router = useRouter();

  const handleAnchorClick = (e: MouseEvent) => {
    if (e.button !== 0) return; // only handle left-clicks
    const targetUrl = (e.currentTarget as HTMLAnchorElement).href;
    const currentUrl = window.location.href;
    if (targetUrl !== currentUrl && window.onbeforeunload) {
      // @ts-ignore
      const res = window.onbeforeunload();
      if (!res) e.preventDefault();
    }
  };

  const addAnchorListeners = () => {
    const anchorElements = document.querySelectorAll('a[href]');
    anchorElements.forEach((anchor) => anchor.addEventListener('click', handleAnchorClick));
  };

  useEffect(() => {
    const mutationObserver = new MutationObserver(addAnchorListeners);
    mutationObserver.observe(document.body, { childList: true, subtree: true });
    addAnchorListeners();

    return () => {
      mutationObserver.disconnect();
      const anchorElements = document.querySelectorAll('a[href]');
      anchorElements.forEach((anchor) => anchor.removeEventListener('click', handleAnchorClick));
    };
  }, []);

  useEffect(() => {
    const beforeUnloadHandler = (e: BeforeUnloadEvent) => {
      e.preventDefault();
      e.returnValue = ''; // required for Chrome
    };

    const handlePopState = (e: PopStateEvent) => {
      if (unsaved) {
        const confirmLeave = window.confirm(
          'You have unsaved changes. Are you sure you want to leave?',
        );
        if (!confirmLeave) {
          e.preventDefault();
          window.history.pushState(null, '', window.location.href);
        }
      }
    };

    if (unsaved) {
      window.addEventListener('beforeunload', beforeUnloadHandler);
      window.addEventListener('popstate', handlePopState);
    } else {
      window.removeEventListener('beforeunload', beforeUnloadHandler);
      window.removeEventListener('popstate', handlePopState);
    }

    return () => {
      window.removeEventListener('beforeunload', beforeUnloadHandler);
      window.removeEventListener('popstate', handlePopState);
    };
  }, [unsaved]);

  useEffect(() => {
    const originalPush = router.push;

    router.push = (url: string, options?: NavigateOptions) => {
      if (unsaved) {
        const confirmLeave = window.confirm(
          'You have unsaved changes. Are you sure you want to leave?',
        );
        if (confirmLeave) originalPush(url, options);
      } else {
        originalPush(url, options);
      }
    };

    return () => {
      router.push = originalPush;
    };
  }, [router, unsaved]);
};

```

Now just pass this component along with our form, filling in the necessary properties:

```ts
  useWarnIfUnsavedChanges(isDirty);
```

**Reply by NyakoFox** [reply]
Is there a way to make this work properly with [next-nprogress-bar](https://github.com/Skyleen77/next-nprogress-bar)? I can't seem to figure out how to modify the code.

The issue is, when you click something on the page that uses their own router.push, it'll show the confirm dialogue, but if you click no, the progress bar still appears. It just doesn't take the user anywhere. I understand that it's because this hook replaces the built-in router.push, but the progress bar uses its own.

**Reply by stychu** [reply]
Nextjs 14.2.14 `popstate` does not trigger when user press browser back button thus navigation via back/forward button is not prevented via this code unfortunately.

There is a hack that makes back button work but it comes with one downside. User need to press back button twice...

```
 if (unsaved) {
	// without this popstate is not even triggered on browser back button..
      window.history.pushState(null, '', window.location.href);

      window.addEventListener('popstate', handlePopState);
      window.addEventListener('beforeunload', beforeUnloadHandler);
    }
```

**Reply by Slightom** [reply]
Thanks @stychu for response! I think you can get rid off this double pressing back button by adding `router.back()` in `else` section in `handlePopState` :)

```
const handlePopState = (e: PopStateEvent) => {
      if (unsaved) {
        if (!window.confirm(unsavedChangesWarningText)) {
          e.preventDefault();
          window.history.pushState(null, '', window.location.href);
        } else {
          router.back();
        }
      }
    };
```

**Reply by stychu** [reply]
Does anyone noticed that when you press back button and `cancel` the back button becomes disabled until u click somewhere in the page ?? Probably due to the pushState but dunno how to fix this. Its bit annoying as for the user he wont understand what to do... Also this happens on chromium. On safari it doesnt even respect the back button and just simply goes back

**Reply by thinhbg2812** [reply]
> Thanks @stychu for response! I think you can get rid off this double pressing back button by adding `router.back()` in `else` section in `handlePopState` :)
> 
> ```
> const handlePopState = (e: PopStateEvent) => {
>       if (unsaved) {
>         if (!window.confirm(unsavedChangesWarningText)) {
>           e.preventDefault();
>           window.history.pushState(null, '', window.location.href);
>         } else {
>           router.back();
>         }
>       }
>     };
> ```

with the else router.back() it's still apear and u have to click it again

**Comment by aphougat** [comment]
I tried something like this and it works, though not for somebody directly changing the url in the browser. 
 
```
'use client'

import {usePathname, useRouter, useSearchParams} from "next/navigation";
import "nprogress/nprogress.css";
import nProgress from "nprogress";
import React, {useEffect, useMemo, useState} from "react";
import ExitWarningOverlay from "../../components/common/ExitWarningOverlay";

export function NavigationEvents({showExitWarningOverlayFeed, setShowExitWarningOverlayFeed}) {
    const pathname = usePathname();
    const searchParams = useSearchParams();
    const [targetUrl, setTargetUrl] = useState(null)
    const router = useRouter();
    const [showExitWarningOverlay, setShowExitWarningOverlay] = useState(false)

    const _push = useMemo(() => router.push.bind(router), []);
    useEffect(() => {
        router.push = (href, options) => {
            if(showExitWarningOverlayFeed){
                setShowExitWarningOverlay(true)
                setTargetUrl(href)
            }else{
                _push(href, options);
            }

        };
    }, [showExitWarningOverlayFeed])

 
    return showExitWarningOverlay ? <ExitWarningOverlay setShowPopup={(status) => {
        if (status) {
            setShowExitWarningOverlayFeed(false);
            setShowExitWarningOverlay(false);
            _push(targetUrl)
        } else {
            setTargetUrl(undefined)
nProgress.done();
            setShowExitWarningOverlay(false);
        }
    }}/> : null;
}
```

and a custom layout which wraps all pages 

```
export const Layout = ({
                           children,
                           classes,
                           message,
                           setMessage,
                           messageClassName = 'alert info',
                           footer = true,
                           showExitWarningOverlayFeed = false,
                           setShowExitWarningOverlayFeed,
                           sx, ...other
                       }) => {
    const {user} = useAuth()
    const settings = useSettingsContext();
    const isNavHorizontal = settings.themeLayout === 'horizontal';
    return (
        <>
            <Suspense fallback={null}>
                <Notifications setMessage={setMessage} message={message} footer={footer}/>
                <NavigationEvents showExitWarningOverlayFeed={showExitWarningOverlayFeed} setShowExitWarningOverlayFeed={setShowExitWarningOverlayFeed} />
            </Suspense>
            <Box sx={{pt: `${isNavHorizontal ? 0 : SPACING}px`, ...sx}} className={classes ? classNames(classes) : null}>

                {children}
            </Box>
            {user ? '' : footer && <Footer/>}
        </>
    )
};
```

**Comment by MartinCura** [comment]
Does someone have a more blessed approach of doing this? Currently i can even navigate away by clicking `Link`s whereas i'd prefer to stop those.
